File: Chapter 05/JS/section_5_6_crypto_tracker/after/infrastructure/adapters/coinGeckoAdapter.js
/**
 * ADAPTER: CoinGecko Implementation.
 * Bridges the Domain's PriceProvider port to the real-world API.
 */
class CoinGeckoAdapter {
    async getBitcoinPrice() {
        try {
            const response = await fetch("https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd");
            
            if (!response.ok) {
                throw new Error("Network response was not ok");
            }

            const data = await response.json();
            return data.bitcoin.usd;
        } catch (error) {
            console.error(`[Adapter] Failed to fetch price: ${error.message}`);
            // Fallback for Archie's demo if the internet is down
            return 65000; 
        }
    }
}

File: Chapter 05/Python/section_5_6_crypto_tracker/after/infrastructure/adapters/coin_gecko_adapter.py
import urllib.request
import json
from core.ports.price_provider_port import PriceProviderPort

class CoinGeckoAdapter(PriceProviderPort):
    """
    ADAPTER 2: The Real Production Adapter.
    Encapsulates all the messy HTTP calls and 3rd-party JSON shapes here.
    """

    def get_bitcoin_price(self) -> float:
        req = urllib.request.Request(
            "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd",
            headers={"User-Agent": "Python App"}
        )

        with urllib.request.urlopen(req) as response:
            json_data = response.read()
            price_data = json.loads(json_data)
            return price_data["bitcoin"]["usd"]